[pavlo.v 최환] 지하철 코드리뷰 부탁드립니다!

choihwan2 opened the pull request:
안녕하세요 파블로입니다!

미션을 진행하던 도중 Section에 대한 관리를 위해서 이리저리 노력하다가 결국 다른 사람들의 힌트와 도움을 받고
코드를 전면 수정하느라 조금 늦었습니다..

예시로 적혀있는 Seciton 과는 다르게 stationId, lineId, distance, SectionType 네가지 만으로 노선 구간을 구현해보았습니다.

JdbcTemplate 을 처음 쓰다보니 dao 쪽 코드가 조금 어지러운 것 같은데.. 보시는데 힘드실까봐 걱정입니다..
잘부탁드리고 강한 피드백 부탁드립니다!!

kakao-pavlo-v committed:
feat : 지하철 역 관리 기능 생성

- StationController 에 StationDao 생성을 static 으로 선언

kakao-pavlo-v committed:
feat : 지하철 역 관리 기능 생성

- StationController 에 StationDao 생성을 static 으로 선언

kakao-pavlo-v committed:
feat : LineController 생성

- 노선 생성 메소드 구현

kakao-pavlo-v committed:
feat : 노선 생성 기능 추가 및 수정

- Line 추가
- LineDao 추가
- 똑같은 노선이름을 가진 노선이 추가될때 400 발생

kakao-pavlo-v committed:
feat : 노선 생성 기능 추가 및 수정

- Line 추가
- LineDao 추가
- 똑같은 노선이름을 가진 노선이 추가될때 400 발생

kakao-pavlo-v committed:
feat : 노선 목록 조회 기능 추가

- showLines 메소드 목록조회 처리
- LineResponse 에 생성자 추가

kakao-pavlo-v committed:
feat : 노선 조회 기능 추가

- showLine 메소드 노선 조회 구현
- LineDao 에 getLine(long id) 메소드 추가

kakao-pavlo-v committed:
feat : 노선 수정 기능 추가

- editLine() 메소드로 기존의 노선 이름과 색깔 변경 구현
- LineDao 에서 editLineById 로 아이디에 해당하는 노선을 찾고 editLine() 메소드 호출

kakao-pavlo-v committed:
feat : 노선 삭제 기능 추가

- deleteLineById 로 노선 삭제 기능 추가
- LineDao 에 deleteLineById() 메소드 추가

kakao-pavlo-v committed:
feat : 노선 삭제 기능 추가

- deleteLineById 로 노선 삭제 기능 추가
- LineDao 에 deleteLineById() 메소드 추가

kakao-pavlo-v committed:
feat : 역 조희,삭제 기능 추가

- StationController 에 있는 showStations() 구현
- StationController 에 있는 deleteStations() 구현

kakao-pavlo-v committed:
feat : 지하철 노선 추가 API 수정

- 상행 종점, 하행 종점과 거리를 Line 객체에 저장

kakao-pavlo-v committed:
feat : 지하철 구간 추가 1차 구현

- Line이 SectionRequest 배열을 가져서 어떠한 구간들을 가지고 있는지 검사
- LineController 에 addSections() 메소드 구현
- SectionRequest 에서 Id값을 가지고 있는지 검사하는 containId 메소드 구현
- 어떠한 방향으로 SectionRequest 가 향하고 있는지 확인 할 수 있도록 SectionType enum 구현

kakao-pavlo-v committed:
feat : 지하철 구간 추가 2차 구현

- Line 에서 SectionRequest 가 아닌 Long(StationId) 을 갖는 배열로 구간을 갖게 만들어봄, distances 로 각 구간 별 거리 구별
- SectionType 에 index 를 추가하여 어느부분에 구간이 추가가 되는지 구현

kakao-eli-nabro committed:
feat : Section, Sections class 생성 및 구간 추가기능 완성

- upDistance, StationId, downDistance 정보를 갖는 section class 생성
- List<Section> 을 갖는 Sections class 생성
- 구간 추가시 예외처리를 하는 matchStation 메서드 추가
- 구간 추가작업을 하는 addSection 메서드 추가

kakao-pavlo-v committed:
feat : 지하철 구간 추가 최종 구현

- Line 에 상행종점과 하행종점을 따로 추가하는 updateTerminalStation 메소드 추가
- LineController 에서의 구간추가 에러시 발생하는 에러코드 수정
- LineResponse 에서 stations 를 가지게 수정
- Section 객체에서 stationId를 갖는게 아닌 Station 객체를 가지고 있게 수정
- SectionType 에 상행종점추가, 하행종점추가 상태 추가

- StationController 에서 findById 하나의 역 검색 메소드 추가
- StationDao 의 메소드들을 모두 static 으로 수정

kakao-pavlo-v committed:
refactor : 노선 구간 추가 리팩토링

- Line 에서의 불필요한 생성자 삭제
- 사용하지 않는 메소드, 주석들 삭제
- Section 생성자 추가

kakao-pavlo-v committed:
feat : 노선 삭제 기능 추가

- LineController 에 deleteSection() 메소드 추가
- Line 에서 역아이디를 찾아 구간을 삭제하는 deleteSection() 메소드 추가
- Sections 에서 예외 처리 및 지우고자 하는 Section index 검색하는 findDeleteSection 메소드 추가

kakao-pavlo-v committed:
refactor : StationDao 를 스프링 빈을 활용하여 관리하게 변경

kakao-pavlo-v committed:
refactor : 노선 관리 코드 스프링 빈 형태로 변환

- LineResponse 기본생성자 추가 (jackson.databind 를 위해)
- Section 에서 Station 객체를 갖는것이 아닌 stationId를 갖게 변경
- LineDao를 static이 아닌 생성자 주입방식으로 변경
- 노선 전체를 조회할시 station 정보는 갖지 않게 수정
- 노선 하나를 조회시 stationDao 에 접근하여 StationResponse List를 생성한 후 LineReponse 를 생성하는 방식으로 변경

kakao-pavlo-v committed:
refactor : 노선 관리 코드 스프링 빈 형태로 변환

- LineResponse 기본생성자 추가 (jackson.databind 를 위해)
- Section 에서 Station 객체를 갖는것이 아닌 stationId를 갖게 변경
- LineDao를 static이 아닌 생성자 주입방식으로 변경
- 노선 전체를 조회할시 station 정보는 갖지 않게 수정
- 노선 하나를 조회시 stationDao 에 접근하여 StationResponse List를 생성한 후 LineReponse 를 생성하는 방식으로 변경

kakao-pavlo-v committed:
refactor : StationDao 에서 Spring JDCB를 활용하게 변경

kakao-pavlo-v committed:
refactor : LineDao Spring JDBC 를 변환

추가해야할 사항 : LineDao 에서 showLineById 구현필요

kakao-pavlo-v committed:
refactor : SectionDao 추가

kakao-pavlo-v committed:
refactor : 구간 추가 메소드 수정 및 로직변경

- Line 이 더이상 Sections 를 갖지 않게 변경
- 노선을 생성할때 초기의 구간들을 생성하는 DB 로직 추가
- Section 이 거리가 아닌 아이디를 갖게 변경
- SectionDao 를 JdbcTemplate 로 수정
- Sections 클래스를 도메인 객체 형태로 변경
- SectionType 이 단순하게 추가되는 구간이 위인지 아래인지 판별하는 enum으로 변경

- SECTION 테이블 속성 변경

kakao-pavlo-v committed:
refactor : Sections 메소드 간결화

- final 과 private 이 빠져있는 부분들 추가
- Sections 의 checkAddSection 메소드에서 타입별로 메소드를 다르게 호출하는 것을 Section이 가진 SectionType으로 해결

kakao-pavlo-v committed:
docs : README.md 수정

sehan commented on README.md:
```diff
@@ -38,16 +38,63 @@ npm run dev
+
+
+
+# 기능 구현 목록
```
꼼꼼하게 잘 작성하셨네요 👍

sehan commented on src/main/java/subway/line/Line.java:
```diff
@@ -0,0 +1,65 @@
+        this.color = color;
+    }
+
+    public Line(LineRequest lineRequest) {
```
Line 은 도메인 객체로 볼 수 있습니다.  layered architecture 측면에서 UI 에 속하는 LineRequest 를 생성자의 인자로 받게 되면 도메인 -> UI 에 의존하게 됩니다. Line 에서 LineRequest 를 사용하는 코드를 제거하여 의존을 제거하고 LineRequest 를 사용해서 Line 을 생성하는 역할을 해주는 객체를 생성해주는 것이 좋겠습니다. 
 보통 그런 객체들을 Factory 라고 부릅니다.

choihwan2 replied:
그렇군요... 팩토리 클래스를 따로 정의해서 구현해보겠습니다!

sehan commented on src/main/java/subway/line/LineDao.java:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    public boolean hasContains(Line line) {
+        String SQL = "SELECT * FROM line";
```
hasContains 를 사용하는 코드를 보면 line 의 이름중복을 체크하는 목적으로 작성된 매소드인거 같습니다. 
Dao 는 데이터의 생성/조회/수정/삭제 하는 역할을 합니다.
그렇다면 Dao 의 역할에 맞게 Dao 는 name 으로 line 을 조회하는 매소드(findByName) 을 제공해주고 Dao 를 사용하는 controller 혹은 service 에서 Line 생성시 검증로직으로 이름이 이미존재하는지 체크하는 매소드(existName) 를 만들어 주면 좋을거 같습니다.

추가로, line 이름으로 이미 존재하는지 여부만 확인하려고 한다면 sql 을 다음과 같이 작성하는것이 좋습니다. 
```
select 1 from line where name = ?
```
전체 테이블조회 쿼리는 정말 필요한 경우가 아니라면 사용은 지양해 주세요

choihwan2 replied:
LineService 객체를 만들어서 수정해 보겠습니다!! 급하게 제출하다보니 dao 객체안에 디비조회 말고 다른 로직들을 사용해버렸네요.. 다양한 시도를 해보고 있는데
```
select 1 from line where name = ?
```
이것과 같은 경우 jdbcTemplate 에서는 어떻게 사용해줘야할까요?? 
queryForObject 메소드 같은 경우 하나의 반환값도 없으면 에러를 출력하게 되는데 이걸 try catch 문으로 감싸서 예외처리를 해줘야하는지, 아니면 이번에 제가 한 방법처럼
```
SELECT count(*) from line where name = ?
```
쿼리문을 이렇게 바꾸고 queryForObject(SQL,Integer.class,name) 이런식으로 작성해 보았는데,  알려주신 방법으로도 적용하고 싶어서 질문드립니다🤔

sehan replied:
count(*) 를 사용하는 것도 좋은 방법입니다.  select 1  대신 count(*) 를 사용하면 Null ( no row ) 가 발생하지 않아서 ResultSet 을 처리하는데 더 이점이 있습니다. 
다만, count 를 사용할때는 count(id) 와 같이 사용하는 것이 좋습니다. count 를 하기 위해 접근하는 column 을 최소화 시키주기도 하고 id 가 index column 일 경우 table 접근없이 더 빠르게 scan 할 수도 있기 때문입니다. 자세한건 대용량데이터베이스 처리 관련 내용을 학습하면 아주 큰 도움이 됩니다. 

select 1 from line where name = ?  은 다수의 row 를 가져오기 때문에 List 타입을 return 하는 query() 를 사용해야 합니다.

choihwan2 replied:
미션이 밀려서 작업하다보니 리팩토링이 늦어지고 있는데 오늘중으로 꼭 PR 보내보도록하겠습니다!! 답변감사합니다

sehan commented on src/main/java/subway/line/LineDao.java:
```diff
@@ -0,0 +1,86 @@
+        String SQL = "SELECT * FROM line";
+        List<Line> lines = jdbcTemplate.query(
+                SQL,
+                (resultSet, rowNum) -> new Line(
```
반복되는 RowMapper 를 따로 빼놓으면 코드중복을 줄여 줄 수 있습니다.

choihwan2 replied:
최고네요!!

sehan commented on src/main/java/subway/line/LineDao.java:
```diff
@@ -0,0 +1,86 @@
+    }
+
+
+    public List<Line> getLines() {
```
목록조회시 네이밍은 보통 findAll() , getAll() 과 같이 많이 사용합니다. 
getLines() 도 나쁜 이름은 아닙니다  하지만 LineDao.getLines() 로 Line 이 많이 등장하기 때문에 그런 측면에서 앞서 제안한 이름들이 더 나은거 같습니다.

choihwan2 replied:
정말 꿀팁이네요.. 감사합니다

sehan commented on src/main/java/subway/line/LineDao.java:
```diff
@@ -0,0 +1,86 @@
+        );
+    }
+
+    public Line getLine(long id) {
```
단건조회시 네이밍도 findById() , getOne() 과 같은 이름으로 많이 사용합니다.

choihwan2 replied:
find로 변경해보겠습니다!☺️

sehan commented on src/main/java/subway/line/LineDao.java:
```diff
@@ -0,0 +1,86 @@
+        );
+    }
+
+    public void editLineById(long id, String name, String color) {
```
Line 수정시 Line 객체를 그대로 받는것이 좋습니다.
Line 의 멤버들이 변경되면 매소드 시그니쳐가 변경되기 때문입니다.  

수정인 경우 네이밍으로 update() 를 사용하거나 save() 를 사용합니다.

choihwan2 replied:
Service 에 메소드를 만들고 
Controller -> Service -> DAO 로 가는 과정과
Line            ->  Line     -> Line   이런식으로 보내준뒤 마지막 쿼리문에서 getId,getName,getColor 이런식으로 접근하는게 좋을까요??

sehan replied:
> Line -> Line -> Line 이런식으로 보내준뒤 마지막 쿼리문에서 getId,getName,getColor 이런식으로 접근하는게 좋을까요??

이게 어떤 의미 인지 잘 모르겠네요. 좀 더 설명해 주면 좋겠습니다.

choihwan2 replied:
처음에 생각했을때 도메인 객체를 계속해서 매개변수로 보내주는 것이 맞는가?? 아니면 처음부터 id,name,color 를 세개를 각각 보낸후에 안에서 로직들을 처리 해주는게 좋은가를 질문하고 싶었던 것이였는데 Factory클래스를 만들면서 Line객체를 생성하고 그것을 보내준후 마지막에 dao 객체에서 get 함수들로 처리를 해보았습니다.

sehan commented on src/main/java/subway/line/LineDao.java:
```diff
@@ -0,0 +1,86 @@
+        String SQL = "INSERT INTO line(name,color) VALUES (?,?)";
+        String SELECT_SQL = "SELECT * FROM line where name = ?";
+        jdbcTemplate.update(SQL, line.getName(), line.getColor());
+        return jdbcTemplate.queryForObject(
```
jdbcTemplate.update() 사용시 KeyHolder 를 사용하거나 
JdbcTemplate 을 감싸서 operation 을 수행할 수 있는 SimpleJdbcInsert 를 사용하며 line 을 다시 조회하기 않고 새로 생성된 Line 의 Id 를 얻어 올 수 있습니다.

sehan commented on src/main/java/subway/section/SectionDao.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    public Section save(Section section) {
+        return save(section.getStationId(), section.getLineId(), section.getPosition());
```
section 테이블의 distance 와 Section 객체의 position 이 같이 data 인데 이름을 다르게 사용하고 있습니다. 
동일한 데이터라면 같은 이름을 사용하는것이 좋습니다. 같은 data 가 여러이름을 가지는 것은 혼동을 주기 때문에 좋지 않습니다.

choihwan2 replied:
리팩토링하는 과정에 빼먹은 부분들이 있었네요.. position으로 통일하는게 좋을것 같아서 통일해보았습니다!

sehan commented on src/main/java/subway/line/LineController.java:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    @PostMapping(value = "/lines", consumes = MediaType.APPLICATION_JSON_VALUE)
+    public ResponseEntity<LineResponse> createLine(@RequestBody LineRequest lineRequest) {
```
Line 생성 시 Line 과 Section 을 생성하고 있고 이 2가지 작업은 하나의 트랜젝션으로 처리되어야 합니다.
두작업을 묶어 "Line 생성" 으로 처리해줄 Service 객체의 필요성이 있습니다. 

Service 를 생성해서 dao 관련 코드를 이동시키고 controller 에서는 request 객체를 Service 로 전달하기 위한 객체변환
Service 에서 return 해준 객체를 response dto 로 객체변환하는 역할을 하도록 분리하면 좋겠습니다.

sehan commented on src/main/java/subway/line/LineController.java:
```diff
@@ -0,0 +1,102 @@
+
+    @DeleteMapping("/lines/{id}")
+    public ResponseEntity<Void> deleteLineById(@PathVariable long id) {
+        lineDao.deleteLineById(id);
```
Line 삭제시 하위 entity 들도 같이 삭제해야 합니다.  하위 entity 로는 section 테이블들이 row 들이 있습니다. 
하나의 트랜젝션으로 동작해야 합니다.

choihwan2 replied:
Section들 객체들도 삭제해줘야하는걸 잊었었네요 추가해보겠습니다!

sehan commented on src/main/java/subway/section/Sections.java:
```diff
@@ -0,0 +1,72 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class Sections {
```
일급컬랙션을 사용해서 Line 에 속한 Section 들에 대한 처리역할들을  잘 옮겨놓으셨네요 👍

sehan commented on src/main/java/subway/section/Sections.java:
```diff
@@ -0,0 +1,72 @@
+        this.sections = sections;
+    }
+
+    public boolean isDuplicatedSection(SectionRequest sectionRequest) {
```
Sections 는 Section 의 일급컬랙션으로 도메인에 속한다고 볼 수 있습니다.
SectionRequest 는 앞서 언급했듯이 UI 에 속한 객체 입니다. 도메인 -> UI 의존이 발생하게 되어 이 부분을 제거하는 것이 좋습니다.

파라미터로 SectionRequest 를 사용하는 매소드에서 의존을 제거해주세요

choihwan2 replied:
제가 제거한 방식이 맞는지 확인 부탁드립니다!

sehan commented on src/main/java/subway/section/Section.java:
```diff
@@ -0,0 +1,84 @@
+    private final long stationId;
+    private final long lineId;
+    private final int position;
+    private SectionType sectionType;
```
sectionType 은 DB 테이블에서 조회해온 Section 객체에서는 사용되지 않고 section 추가 과정에서 만들어진 임시객체(not persist) 에서만 사용되고 있습니다. 이것은 Section 객체가  2가지 책임을 가지고 있다고 볼 수 있습니다.  즉, SRP 원칙을 위배하고 있습니다.

id, stationId, lineId, position 과 함께 section entity 의 역할 , sectionType 에 따른 section util(?) 의 역할
두 가지를 분리하는 것이 좋겠습니다. 

Sections.checkAddSection() 에서 standardSection 이 하는 역할을 Section 에서 분리해주세요

choihwan2 replied:
꽤 많은 시간을 고민하면서 분리를 해보려했는데 잘 되지 못한 것 같습니다.. Service쪽으로 로직들이 많이 이동한 것 같은데 보시고 피드백 부탁드립니다.

sehan reviewed (changes requested):
안녕하세요. 지하철 미션 리뷰를 맡게된 니오입니다. ☺️

기능구현을 좀(?) 다른방식으로 하셨네요 😅
미션의 의도를 이렇게 돌아가도 되는건지는 모르겠습니다만 distance 를 position 으로 바꿔놓은 점은 신선했습니다. 하지만 section 을 추가할때 두 역간의 거리를 distance 로 받고 section table 의 distance 에는 출발역에서의 절대적인 거리를 저장하고 있어서 혼돈이 좀 있었습니다.

그 외 리팩토링을 진행할 부분들이 보여서 코멘트를 남겨놨습니다. 
확인 부탁드려요

수정요청 드리겠습니다.

sehan commented:
> 예시로 적혀있는 Seciton 과는 다르게 stationId, lineId, distance, SectionType 네가지 만으로 노선 구간을 구현해보았습니다.

다른 구현체들과 풀이가 달라서 로직을 확인하는데 시간이 좀 걸렸네요 😂
 
> JdbcTemplate 을 처음 쓰다보니 dao 쪽 코드가 조금 어지러운 것 같은데.. 보시는데 힘드실까봐 걱정입니다..
> 잘부탁드리고 강한 피드백 부탁드립니다!!

Dao 코드 어려움 없이 잘 봤습니다. 😊

kakao-pavlo-v committed:
refactor : 패키지 생성과 Service 클래스 생성

- LineService, SectionDao 생성, 비즈니스 로직 추가
- 전체 Exception을 다루는 GlobalControllerAdvice 클래스 추가
- 패키지 세분화

kakao-pavlo-v committed:
refactor : Service, Exception 리팩토링

- StationResponse 에 of 정적팩토리 메소드 추가
- AlreadyExistsException 을 만들어 비슷한 역활의 Exception 상속으로 변환
- Section 이 가지고 있던 SectionType 삭제
- StationService 추가 및 Controller에 있던 로직 Service로 이동

- test INTERNAL_SERVER_ERROR => BAD_REQUEST 로 변경

kakao-pavlo-v committed:
refactor : keyHolder 사용

- LineDao, SectionDao 에서 KeyHolder 를 사용해 save 후 객체 받아오는 메소드 변경

kakao-pavlo-v committed:
refactor : LineFactory 메소드 추가

choihwan2 commented:
너무 늦어서 죄송합니다.. 새로운 미션을 진행하면서 알게 된 것들과 패키지별로 나누면서 Service 객체들을 새롭게 다 생성하고 하면서 많이 늦어졌네요.. 기존의 방법과는 다르게 미션을 진행하다보니 이걸 어떻게 하면 좀 더 효율적으로 바꿀수 있을까 너무 고민을 하다보니 더 늦어지고 점점 더 마음에 들지 않는 코드들이 생성되는 것 같았습니다. 피드백 주신 내용들 빠짐없이 해봤는데, 혹시라도 빠지거나 별로인 내용 있으면 다시 피드백주시면 감사하겠습니다!

sehan commented on src/main/java/subway/line/application/LineService.java:
```diff
@@ -0,0 +1,79 @@
+    @Transactional
+    public void delete(long id) {
+        lineDao.deleteLineById(id);
+        List<Section> list = sectionDao.getSections(id);
```
Line 이라는 객체를 삭제하는 책임을 도메인 영역으로 설계하지 않았기 때문에 삭제에 대한 방식은 위 코드와 같이 조회 후 하나씩 삭제를 하거나 line 에 해당하는 section 을 한번에 모두 삭제하는 형태로 구현할 수 있습니다. 이 책임은 온전히 sectionDao 에게 넘어 가는 것입니다.  
 그런 관점에서 line 에 포함된 section 들을 삭제하는 방식을 구현한다면 각 Dao 에게 삭제요청을 하는 것이 더 효과적입니다.
```
sectionDao.deleteByLineId(lineId)
lineDao.deleteById(lineId) 
```

sehan commented on src/main/java/subway/line/domain/Line.java:
```diff
@@ -0,0 +1,63 @@
+
+import java.util.Objects;
+
+public class Line {
```
현재 Line 은 도메인 객체로 가지고 있는 역할이 id, name, color 의 getter 입니다. 
section 추가/삭제 같은 책임을 가지도록 한다면 더 좋은 객체설계가 될 수 있습니다.

sehan commented on src/main/java/subway/section/application/SectionService.java:
```diff
@@ -0,0 +1,48 @@
+        this.sectionDao = sectionDao;
+    }
+
+    public void insertSection(SectionRequest sectionRequest, long lineId) {
```
Section 추가를 위한 비지니스 로직이 아직 남아 있습니다. 

로직을 도메인으로 이동시켜준다면 아래와 같은 코드형태가 되어야 합니다. 
```
sections.addSection(new Section())
sectionDao.deleteByLineId(lineId)
sectionDao.saveAll(sections.getSections())
```

현재 지하철 서비스 미션도 진행중 일거라 생각됩니다. 
지하철 서비스 미션의 코드도 함께 참고하면서 도메인으로 비지니스 로직을 이동시키는 방법에 대해서 고민해 보면 좋겠습니다. ☺️

sehan commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,23 @@
+    primary key(id)
+    );
+
+create table if not exists SECTION
```
SECTION 테이블은 두역간의 정보를 담는 용도로 설계된 entity 입니다. 두역에 의존적인 distance 가 주요 field 였는데 line 을 중심으로 역의 순서(position) 를 지정하는 방식으로 변경을 하셨네요. 😅

테이블 명을 변경해야 할듯 합니다. SECTION 보다는 LINE_STATION 이나 STATION_POSITION 이 필드의 특성을 잘 표현해 줄 거 같네요.

sehan reviewed (approved):
기존의 방법과 다른 풀이로 접근하다 보니 Line , Section , Station 도메인에 대입하는게 더 어색하고 이상했을거 같네요.  Section 은 역간 정보를 담아내는 객체인데 DB 에 저장데이터는 역간 정보가 아닌 Line 에서 역의 위치정보를 담는 방식으로 하다 보니 Section 이라는 도메인이 가지는 역할이 이름과는 다른 방식으로 구현되어 혼돈을 준 것이 아닌가 싶습니다. 

지하철 서비스의 패키지 구성에 영향을 받아 패키지 구성을 재구성 하셨습니다. Dao 에서 도메인 객체를 매핑하는 부분도 참고해서 개선시키면 Service 에서 비지니스 로직을 제거하는데 도움이 될 것입니다. 

이번 PR 은 여기서 merge 하도록 하겠습니다. 

덧, 비지니스 로직 제거에 관련한 코멘트를 남겨 두었습니다. 참고 하시고 이후 미션 진행하면서 함께 고민해주세요. 고생하셨습니다. ☺️
